packages/angular_devkit/schematics/tools/workflow/node-workflow.ts

```diff
@@ -0,0 +1,133 @@
+import { _throw } from 'rxjs/observable/throw';
+import { concat, concatMap, ignoreElements, map } from 'rxjs/operators';
+import { NodeModulesEngineHost, validateOptionsWithSchema } from '..';
+import { DryRunEvent } from '../../src/sink/dryrun';
+import { BuiltinTaskExecutor } from '../../tasks/node';
+
+export class NodeWorkflow implements workflow.Workflow {
+  protected _engine: SchematicEngine<{}, {}>;
+  protected _engineHost: NodeModulesEngineHost;
+
+  protected _dryRunSink: DryRunSink;
+  protected _fsSink: Sink;
+
+  protected _context: workflow.WorkflowExecutionContext[];
+
+  constructor(
+    protected _host: virtualFs.Host,
+    protected _options: {
+      force?: boolean;
+      dryRun?: boolean;
+    },
+  ) {
+    /**
+     * Create the SchematicEngine, which is used by the Schematic library as callbacks to load a
+     * Collection or a Schematic.
+     */
+    this._engineHost = new NodeModulesEngineHost();
+    this._engine = new SchematicEngine(this._engineHost);
+
+    // Add support for schemaJson.
+    const registry = new schema.CoreSchemaRegistry(formats.standardFormats);
+    this._engineHost.registerOptionsTransform(validateOptionsWithSchema(registry));
+
+    this._engineHost.registerTaskExecutor(BuiltinTaskExecutor.NodePackage);
+    this._engineHost.registerTaskExecutor(BuiltinTaskExecutor.RepositoryInitializer);
+
+    // We need two sinks if we want to output what will happen, and actually do the work.
+    // Note that fsSink is technically not used if `--dry-run` is passed, but creating the Sink
+    // does not have any side effect.
+    this._dryRunSink = new DryRunSink(this._host, this._options.force);
+    this._fsSink = new HostSink(this._host, this._options.force);
+
+    this._context = [];
+  }
+
+  get context(): Readonly<workflow.WorkflowExecutionContext> {
+    const maybeContext = this._context[this._context.length - 1];
+    if (!maybeContext) {
+      throw new Error('Cannot get context when workflow is not executing...');
+    }
+
+    return maybeContext;
+  }
+  get reporter(): Observable<DryRunEvent> {
+    return this._dryRunSink.reporter;
+  }
+
+  execute(
+    options: Partial<workflow.WorkflowExecutionContext> & workflow.RequiredWorkflowExecutionContext,
+  ): Observable<void> {
+    /** Create the collection and the schematic. */
+    const collection = this._engine.createCollection(options.collection);
+    const schematic = collection.createSchematic(options.schematic);
+
+    let error = false;
+    const dryRunSubscriber = this._dryRunSink.reporter.subscribe(event => {
+      error = error || (event.kind == 'error');
+    });
+
+    const parentContext = this._context[this._context.length - 1];
+    const context = {
+      ...options,
+      debug: options.debug || false,
+      logger: options.logger || new logging.NullLogger(),
+      parentContext,
+    };
+    this._context.push(context);
+
+    return schematic.call(options.options, of(new HostTree(this._host)), {
```

Comment:
What do you think about inverting the responsibility here and having the engine `execute` the schematic.  This would give the engine access to the host and even the sinks as they could be passed in as a  _ tree sink_ Observable parameter.  So the passed in sinks would be an observable chain that would get access to the event stream and then either act on them, drop them, pass them along, or some combination.

Reply:
It's an interesting proposal. I think it could definitely work. I've thought about having the workflow as the place you could add sinks, but there's more implications than just that.

In general I'd say that's post 6.0 work. Having the Workflow abstraction would allow us to refactor this anyway.

Reply:
Yes. I agree. It was more of a brainstorming comment then anything else.